cesql/spec.md

```diff
@@ -327,26 +362,47 @@ The following tables show the built-in functions that MUST be supported by a CES
+| `SUBSTRING(x, pos): String x Integer x Integer -> String`  | Returns the substring of `x` starting from index `pos` (included) up to the end of `x`. Characters' index starts from `1`. If `pos` is negative, the beginning of the substring is `pos` characters from the end of the string. If `pos` is 0, then returns the empty string. Returns the empty string and a _FunctionEvaluationError_ if `pos > LENGTH(x) OR pos < -LENGTH(x)`. |
+| `SUBSTRING(x, pos, len): String x Integer x Integer -> String` | Returns the substring of `x` starting from index `pos` (included) of length `len`. Characters' index starts from `1`. If `pos` is negative, the beginning of the substring is `pos` characters from the end of the string. If `pos` is 0, then returns the empty string. If `len` is greater than the maximum substring starting at `pos`, then return the maximum substring. Returns the empty string and a _FunctionEvaluationError_ if `pos > LENGTH(x) OR pos < -LENGTH(x)` or if `len` is negative. |
 
 #### 3.5.2. Built-in Math functions
 
 | Definition                                                 | Semantics                                                                                                                                                                                                                  |
 | ---------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
-| `ABS(x): Integer -> Integer` | Returns the absolute value of `x` |
+| `ABS(x): Integer -> Integer` | Returns the absolute value of `x`. If the value of `x` is `-2147483648` (the most negative 32 bit integer value possible), then this returns `2147483647` as well as a _MathError_. |
+
+#### 3.5.3 Function Errors
+
+As specified in 3.3, in the event of an error a function MUST still return a valid return value for it's defined return type.
+A CESQL engine MUST guarantee that all built-in functions comply with this. For user defined functions, if they return one or more errors
+and fail to provide a valid return value for their return type the CESQL engine MUST return the zero value for the return type of the
+function, along with a _FunctionEvaluationError_.
 
 ### 3.6. Evaluation of the expression
 
-Operators MUST be evaluated in order, where the parenthesized expressions have the highest priority over all the other
-operators.
+Operators and functions MUST be evaluated in order of precedence, and MUST be evaluated left to right when the precedence is equal. The order of precedence is as follows:
+
+1. Function invocations
+1. Unary operators
+  1. NOT unary operator
+  1. `-` unary operator
```

Comment:
I don't think you got the indenting you were hoping for, I see this:
<img width="440" alt="image" src="https://github.com/cloudevents/spec/assets/1944671/fc84a9d8-23c8-4288-9b88-8c78f5c640f4">

and this new direction is much clearer than what you had before.